### DISCUSSION TITLE
Data storage instead of updates for user scripts

### DISCUSSION THREAD

**Discussion by xarantolus** [bromite/bromite#2113] [Ideas]
Hi, I have questions/ideas regarding user scripts that want to share/keep state between different domains.

I've created two [user scripts](https://github.com/xarantolus/bromite-userscripts) that can basically be split into a "data" and a "code" part. The code part doesn't change between releases, but the data needs to be updated from time to time for the scripts to work well.

As of now, it is kind of tiresome to download the scripts each week manually and install them, so some kind of update mechanism would be nice. #792 included plans on self-updating scripts, but I don't think this should be implemented (or only in a very limited way, honestly don't know?) due to security problems that arise when the author of a widely-installed script turns on "evil mode" and updates the script without users knowledge (honestly: who actually reads the code of an 5MB script? nobody, users would just click "update" as they've done with so much other stuff).

So when trying to get around updating the script itself but still updating data used by the script (which could also be nice from a security perspective, as now the potential for doing malicious stuff would be reduced), I realized that some kind of `localStorage`-like data storage would be nice to have for userscripts. That way, the userscript could be updated using a special website that passes the new data to the script, and then the script could use that data on any domain (This site could of course also notify the user if their script itself needs to be updated, e.g. in case a new type of rule is supported).

So the old "update" flow would be this:
1. Remember that you should probably update the script
2. Go to the releases of that script and download it
3. Install the script file

And the new version would look like this:
1. The script is injected on a website and checks if the stored data (e.g. content blocker rules) is recent
2. If not, the script just opens a website that provides an update mechanism
3. The script provides some kind of callback (only to that specific page) so the "update" site can pass data to the script 
4. The script can now use that updated data on any other page visits, as the data storage is cross-domain, specific to this script.

This second mechanism is somewhat similar to what would be possible with functions like `GM_setValue` and `GM_getValue` [that Chromium doesn't appear to support](https://www.chromium.org/developers/design-documents/user-scripts/), so basically my question boils down to the following: **Is there any way a script could use/implement cross-domain storage of data?**

I think this approach coupled with not allowing `eval` for user scripts (as otherwise updating data is just a more fancy way of updating code) could provide a somewhat secure (?) compromise between "updating the whole script, possibly introducing malicious code" and "not updating script data at all".

**Comment by csagan5** [comment]
> honestly: who actually reads the code of an 5MB script? nobody, users would just click "update" as they've done with so much other stuff

The fact that users are responsible for that cannot be dismissed with some hand-waving: they are. Only guarantee in Bromite is that the code that you have already downloaded will be executed, nothing else downloaded from an URL. Also: there is not going to be anyone reviewing the user scripts for anyone else.

> Is there any way a script could use/implement cross-domain storage of data?

Do other browsers (e.g. Firefox) have support for this?

**Reply by xarantolus** [reply]
> Do other browsers (e.g. Firefox) have support for this?

The way other userscript managers handle this are the mentioned `GM_setValue` and `GM_getValue` functions that store data per-script and can be accessed cross-domain by that script. This can be useful to e.g. store settings for a script.

But for my use-case they could be an easier way to update the scripts' data, with limited capabilities only. Full script updates probably make more sense than the method I suggested though.

**Reply by csagan5** [reply]
Yes, I also think that `GM_setValue` would not be the right API to store something which is not configuration e.g. large quantity of data.

So if I understand correctly there are no other known implementations of this (a complex user script would become an extension in the use-case of other browsers) and any feature added for this would be Bromite-specific.

I am not in favour of that, it would become a slippery slope to implement some sort of custom Bromite extensions and it definitely should not be added based on the use-case of a single user script.

If you want to implement support for this in your custom Chromium fork the best approach is probably to declare "assets" in the user script and then have the browser keep them up-to-date, but as explained I do not wish to accept patches for this functionality.

**Reply by uazo** [reply]
> but as explained I do not wish to accept patches for this functionality

the alternative could be to allow the user to select the userscripts folder, which at that point could be in any position and can be updated with external tools. when loading, bromite will check the hash of the content, and if different from the last one, deactivate the userscript and communicate it to the user who can decide what to do

**Reply by csagan5** [reply]
> > but as explained I do not wish to accept patches for this functionality
> 
> the alternative could be to allow the user to select the userscripts folder, which at that point could be in any position and can be updated with external tools. when loading, bromite will check the hash of the content, and if different from the last one, deactivate the userscript and communicate it to the user who can decide what to do

This would break the current functionality of downloading & using scripts directly?
This proposal is clever but also does not keep in consideration the importance of users reading the source: imagine that a lot of people are using a heavily customised Bromite with an external userscripts folder updated via another app, what I expect to happen is:
* more bug reports about problems introduced by user scripts (this is already happening)
* users less conscious about what they are running on each browser navigation ("I only selected this folder explained in this site here")
* if there is a sufficiently popular user script author that gets compromised then a lot of Bromite users will be affected and it will definitely create the name of an "unsecure browser" for it
* at that point Bromite would have to start blacklisting user scripts or remove the functionality altogether, even if an initial user interaction is required to enable user scripts (either in current or in a future approach)

For all these reasons I do not accept patches to make the user scripts feature more advanced but also more Bromite-specific. And as mentioned before nothing should be done only based on the use-case of 1 user script.

**Reply by uazo** [reply]
> This would break the current functionality of downloading & using scripts directly?

no, theoretically.

> what I expect to happen is

sure, it is a possible scenario.
but bromite is not "secure in all circumstances", nor will it ever be, it is always the user who makes the difference. to date, everything is possible in a rooted device.

> And as mentioned before nothing should be done only based on the use-case of 1 user script.

yes, we're just talking, to me userscripts are fine as they are.
